### DISCUSSION TITLE
MSI installer should be able to upgrade to new minor versions within the same major version

### DISCUSSION THREAD

**Discussion by MaanuelMM** [microsoft/openjdk#37] [Q&A] [answered]
When upgrading Microsoft Build of OpenJDK from 11.0.10+9 to 11.0.11+9 using the MSI installer, previous version isn't uninstalled and replaced by the new one. Instead, both versions remain installed.

For example, the AdoptOpenJDK installer is able to upgrade by removing the old version and keeping the new one within the same major JDK version, as described on their [website](https://adoptopenjdk.net/installation.html) (the same is true for [Azul](https://docs.azul.com/zulu/zuludocs/ZuluUserGuide/UpgradingZulu/UpgradeZuluOnWindows.htm)).

I don't know if this is a bug or something intentional, but I do think it would be a very valuable feature.

![image](https://user-images.githubusercontent.com/45206520/115967547-7c2a6380-a533-11eb-91a2-e860e630bd77.png)

**Comment by karianna** [comment]
@MaanuelMM - Thanks for your feedback - our team will take a look at this internally!

@gdams - Are you able to compare / contrast vs Adopot's installer (is there a user option to choose install alongside or upgrade)?

**Reply by MaanuelMM** [reply]
It's kinda the same installer, but instead of suggesting me to choose the features I want to install, it immediately takes the ones from the previous installation (which I can modify, of course).

And it also automatically uninstalls previous version as the progress bar says. No warning, no alert, no option. Mandatory and automatic.

**Reply by karianna** [reply]
See #123 which partly tracks a similar concept

**Reply by si-chan** [reply]
I feel this is somewhat related to the fact that there is no separate JRE any more.
The previous behaviour with JRE 8 (e.g. Amazon Corretto) was that it would install in a jre8 folder and any minor update would simply overwrite what was there.
With the JDK 8 though, you would get side-by-side folder versions - from a development point of view, that's useful if you are regression testing your program and want to compare.

So, ideally there should be options for either behaviour.
That said, I would hazard a guess that there are far, far more end-user installation of the current LTS JDK than developer installs... at least I'd hope so!

**Reply by gdams** [reply]
at a guess, I'd imagine it's because we (Microsoft) are not setting the `UPGRADE_CODE_SEED` when we build the installers. I'll take a look now...

**Reply by gdams** [reply]
@karianna I have prepared a fix internally, hopefully, we will have rolled this out for the July PSU release. It's worth noting that the new MSI installer won't be able to upgrade the previous MSI's that were built without the fix so we will have to wait until the September PSU to confirm that the fix successfully upgrades the July PSU.

**Reply by brunoborges** [reply]
@gdams before we roll this out, let's make sure we have good UX testing. Can you please prepare a demo for our internal team?

**Comment by jax7778** [comment]
One additional Vote for a switch and/or the default behavior for the MSI to update minor revisions. It would make updating and securing installations far easier if we could just push the latest msi when it comes to it. I would honestly like it added to WSUS, but I think that is a pipe dream, but an install switch to have the MSI update instead of install alongside seems reasonable.

**Comment by karianna** [comment, marked as answer]
This has been resolved.
